### PR TITLE
[fast-client][test] batchGet validation for Fast Client

### PR DESCRIPTION
**Pull request by m-nagarajan**
## Summary
Validating batchGet for fast client by cleaning up and adding more tests.

1. Moved all the setup for the tests to [AbstractClientEndToEndSetup.java](https://github.com/linkedin/venice/pull/117/files#diff-d12386de89800486959f6f71f4cd35da9727c0e7a5029f23c4b35c58e368408d) and cleaned up [AvroStoreClientEndToEndTest.java](https://github.com/linkedin/venice/pull/117/files#diff-fca448884424f1f7bf4acd704116767ae0302004fee566987448b31bfc6cb36c) and [BatchGetAvroStoreClientTest.java](https://github.com/linkedin/venice/pull/117/files#diff-1c203ea7d8742d684419f62d6b9bb83d2a387a506dc8edc406f2565bda077224) to use this setup.
2. Test for all combinations of `useDaVinciClientBasedMetadata`, `multiGet`, `dualRead`, `speculativeQueryEnabled` and multiple batch sizes. This applies to Gzip and zstd compression as well.
3. Added tests in [R2TransportClientTest.java](https://github.com/linkedin/venice/pull/117/files#diff-ae4b57797fa18ce1d4483059b16a3ddce303b63742ce549896097ee11ba35195) and [HttpClientBasedR2ClientTest.java](https://github.com/linkedin/venice/pull/117/files#diff-9111265b9932aeba8b1347b8981f30a2b7d810e567f34fdeae78f900b307118b) 
4. Modified [RetriableAvroGenericStoreClientTest.java](https://github.com/linkedin/venice/pull/117/files#diff-a16fba0d1d07bf0dc26e410b59a5367e4e8b1fe08449163fd3a3210e2c1f80af) to use `StatsAvroGenericStoreClient` to wrap around `RetriableAvroGenericStoreClient` to test for stats as well.
5. fixed some flaky issues
6. Added some comments on the flow and some questions in TODO.

TODO for this project (Will be continued in a different PR):
1. Add tests for batchGet in `RetriableAvroGenericStoreClient` and `DualReadAvroGenericStoreClientTest`
2. More coverage for stats: Add/modify tests to record all the stats.
3. Check if some more cases can be added in `BatchGetAvroStoreClientUnitTest`
4. Fix `batchGet()` looping through through single `get()` by calling `streamingBatchGet()` => but as `batchGet()` API is being used in PROD, this change can go in only after proper testing of `streamingBatchGet()`. This change was added in this PR but later reverted using this [commit](https://github.com/linkedin/venice/pull/117/commits/689d7b14765c41e7e209602db9102054090255c9) to make this PR mergeable and can continue adding more tests in a new PR and finally when `streamingBatchGet()` is validated, can bring in this change.

## How was this PR tested?
Internal CI

## Does this PR introduce any user-facing changes?
- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:315]: Thanks for fixing other typos, and this one also a typo:
longTailRetryThresholdFor**Batcht**GetInMicroSeconds
Can you fix it as well?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:299]: This is a good question.
Today, the batch-get size enforcement is handled in Venice Router and it is configurable in store-level.
In Fast-Client, it is still an open question about how to setup the batch-get limit or whether we need any limit at all.
At the first step, It is better to keep the similar semantics, but enforced it in fast-client and later, we can adjust or remove this limit after a thorough validation.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java:48]: Is this link accessible permanently?
If not, we should remove it and if you are interested, maybe we could document some ASCII-based workflow, and in the high-level, I feel the above javadoc is clear enough.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java:29]: Why remove this?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:145]: Check this:
https://github.com/linkedin/venice/blob/main/clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/LeastLoadedClientRoutingStrategy.java#L56

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:178]:
```
TODO what if there are only 2 replicas and both the replicas are in the same server?
```
This is not possible.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:203]: This is handled by the following `allOf` to wait for all the futures.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:248]: Do we need this level of distinction?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DualReadAvroGenericStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DualReadAvroGenericStoreClient.java:131]: Can you elaborate on this issue?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:87]: I suggest to remove all the `QQ` from the javadoc, which could be misleading for other committers.
Also the retry volume limit is meant to reduce the total allowed retry per node.
Today, the retry volume can be 1 retry request per the original request at max, which is unexpected, and we would like to design some mechanism to make it configurable, such as retry at most the slowest 5% of traffic, otherwise, too many retry requests could cause cascading failure.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java:49]: The intention is to hit the implementation details of `getReplicas`, so that the implementor can decide whether the original route will be excluded or not.
For example, the underlying implementation may return more replicas than asked for other purposes, and you can check the `ClientRoutingStrategy` impl.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java:70]: Can you suggest a better name? `trackRequestToInstance`?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java:111]: I remembered the instance will be treated as unhealthy right away and it will be reseted after the configured timeout.

**Review by gaojieliu — COMMENTED** [review]
Sorry for late review, and I don't think I can follow up with your replies next week, and I hope these comments could help you make some progress.

**Comment by asaxena76** [comment]
Thanks for the effort and sorry for the late review. One overall comment. Please squash the commits into 1 or 2 logical commits. This should be done before merging for sure.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java
Comment (asaxena76) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java:48]: You could attach a digram to the PR as well and in github docs.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java
Comment (asaxena76) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:299]: Also I don't think we should allow unlimited number of keys in a request. There should be a (configurable) upper limit. We aren't truly streaming at this point since we ask for all the keys in the request itself. Hence a limit will need to be applied or we will run into memory issues.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java
Comment (asaxena76) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:87]: Discussion and questions on the code should be moved to comments in the PR not in the code. Like gaojie said the QQ comments will need to be removed before merge.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java
Comment (asaxena76) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java:49]: ` QQ based on above TODO: Looks like the retry is being scheduled once via scheduler (LONG_TAIL_RETRY) and
If the intention is not to optimize for the first try and not for retries, have a couple of questions on the implementation:
   *  1. Calling getReplicas() for every retry => Is this necessary to get the updated metadata and not have the old stale metadata?
   *  2. loop and check whether some replica is already tried => is this okay as the number of replicas is very small?
`
For 1 , I guess its not necessary but why wouldn't you. We do want to get the latest metadata since its possible that new replicas got added between the requests. 
For 2 Its just filtering to remove the excluded instances. it won't be a performance issue since the number of replicas is under is small

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java]: Fixed it.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java]: Below items needs to be taken care in a new PR
1. the variable names for this limit is different for routers config and fast client. `maxKeyCountInMultiGetReq` vs `maxAllowedKeyCntInBatchGetReq`. Clean it up.
2. change it to behave like thinclient to start with: Set default to `500` as a global limit to start with and probably use per store limit as well.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java:48]: Yes, It's accessible permanently. I attached the link as this can be improved continuously.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DelegatingAvroStoreClient.java]: Thanks for finding it. Fixed it, should have been removed by mistake.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:145]: That part of the code talks about the case: `requiredReplicaCount < availReplicas.size()`, my question was about the case: `requiredReplicaCount > availReplicas.size()`. I think these won't be an issue as we ideally don't set `requiredReplicaCount` more than 2 (original query + speculative query). But if 2 replicas out of 3 are down, this situation will be hit before helix comes in and brings up the 2nd and 3rd replicas.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java]: Is this because Helix takes care of this?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java]: Makes sense. Thanks.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java]: we do log the exceptions leading to this errors, so we should be okay. Resolving this.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DualReadAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DualReadAvroGenericStoreClient.java:131]: Slack [thread](https://linkedin-randd.slack.com/archives/G0GEB715Y/p1670440026847499) on this issue.
Filed a bug for the same: [VENG-10197](https://jira01.corp.linkedin.com:8443/browse/VENG-10197).

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java]: Makes sense thanks. Yes, these were added as a conversation starter for this PR, will remove this in the next iteration.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java]: Makes sense. Thanks.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java]: I think we have 2 concepts here: Blocked instances(which are not retried until the open requests is served or times out) and unhealthy instances(being retried continuously until it becomes a blocked instance). This case will implicitly fall under blocked instances as there are too many requests waiting on them, so not marked unhealthy. This is my understanding after looking at the code again.

**Comment by m-nagarajan** [comment]
Thanks @gaojieliu and @asaxena76 for the reviews. I have cleaned up my comments in the code based on this.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java]: renamed it to `trackHealthBasedOnRequestToInstance`. What do you think?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (asaxena76) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:145]: The request should fail if we don't have required replica count. If we are using speculativequery then we need to have 2 replicas available

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (asaxena76) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java]: I think the idea was this - If the status is 429 then it means request quota was exceeded . However the instance itself is not unhealthy . Perhaps the client catches the exception and  slows down the rate of requests so the next one should go through. You could also argue that we should aggressively limit the client from sending more requests but that would be new logic . Don't think we need to go there right now. The instance should still be treated healthy for this error code

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroGenericStoreClient.java
Comment (asaxena76) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroGenericStoreClient.java:56]: Yes they have different guarantees. batchGet will block and return only when all key values have been retrieved

**Review by asaxena76 — APPROVED** [review]
Thanks for all the refactoring and improved comments. General comment is to keep changes small in next PR and logically connected.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java
Comment (FelixGV) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:145]: Speculative query should be best effort. If we have 1 replica available and speculative query enabled, then we should still execute that 1 request, not fail.

Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java:145]: Thanks @FelixGV. It's currently working that way and so resolving this.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DualReadAvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DualReadAvroGenericStoreClient.java]: Resolving this, to be taken care in the future PR.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java]: Resolving this. We can revisit if needed.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/InstanceHealthMonitor.java]: Makes sense. Modified the comment to reflect the same. Thanks.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroGenericStoreClient.java
Comment (m-nagarajan) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AvroGenericStoreClient.java]: But I think the plan is to default the batchGet() call to use streamingBatchGet() internally, similar to thin client. Marking this as resolved and will be revisited when making this actual API change.

File: clients/venice-client/src/test/java/com/linkedin/venice/fastclient/utils/TestClientSimulator.java
Comment (m-nagarajan) [reply clients/venice-client/src/test/java/com/linkedin/venice/fastclient/utils/TestClientSimulator.java]: But should we not cleanup but only have it disabled by default and can be enabled as fallback for the initial ramp up of FC batchGet and the code can be removed only if the initial deployments are stable. Resolving this now as this is not in the scope for this PR. CC: @gaojieliu @xunyin8

File: clients/venice-client/src/test/java/com/linkedin/venice/fastclient/utils/TestClientSimulator.java
Comment (m-nagarajan) [reply clients/venice-client/src/test/java/com/linkedin/venice/fastclient/utils/TestClientSimulator.java:436]: Thanks for the thoughts on this. Resolving as not in scope of this PR.

**Pull request merged**
